### DISCUSSION TITLE
Clipped Design Vector Objects Become Visible in New PDF When Redrawing

### DISCUSSION THREAD

**Discussion by Danishkhilji** [pymupdf/PyMuPDF#3977] [Looking for help]
Hello, I’m working with a PDF that contains vector design objects, some of which are clipped, meaning parts of the design are intentionally hidden in the original PDF. I’m attempting to extract and redraw these objects into a new PDF file.

I’m using the get_drawings() method from PyMuPDF (MuPDF) to extract the vector paths and redraw them. However, when I draw these extracted paths onto the new PDF, the clipped areas that were originally hidden in the source PDF are now visible, and the design doesn’t match the original.

**Steps I’m Taking:**
I extract vector drawings using the page.get_drawings(extended=True) method.
I then iterate over the paths and redraw them on a new PDF page using the Shape drawing methods (draw_line, draw_quad, draw_rect, etc.).
The problem is that when I do this, the previously clipped areas of the vectors are now visible, which is not the desired result.

**Actual Behavior:**
The previously clipped portions of the vector objects are now fully visible, resulting in an inaccurate reproduction of the original design.

How can I correctly handle the clipping regions when redrawing the vector objects so that the clipped areas remain hidden in the new PDF? 

Any guidance or solutions would be greatly appreciated. Thank you!

**here is the code.**

`
@app.post("/pdf-parsing")
async def parse_pdf(file: UploadFile = File(...)):
    try:
        pdf_bytes = await file.read()

        pdf_document = fitz.open(stream=pdf_bytes, filetype="pdf")

        page = pdf_document.load_page(0)

        # Define the coordinates of the area need to clip 
        x0, y0, x1, y1 = 1122.5198, 90.7085, 1547.7166, 856.0625
        clip_rect = fitz.Rect(x0, y0, x1, y1)

        # Create a new PDF with the same page size
        new_pdf = fitz.open()
        new_page = new_pdf.new_page(width=page.rect.width, height=page.rect.height)

        # Extract only the "Design" layer
        ocg_layers = pdf_document.get_ocgs()
        design_layer = None

        # Loop through layers and find the "Design" layer
        for layer_id, layer_info in ocg_layers.items():
            if layer_info["name"] == "Design":
                design_layer = layer_info
                break

        if not design_layer:
            raise HTTPException(status_code=404, detail="Design layer not found in the PDF.")

        # Create a shape to draw the new content
        shape = new_page.new_shape()

        # Get all vector drawings from the page
        paths = page.get_drawings(extended=True)

        for path in paths:
            if path.get("layer") == "Design":  
                if "rect" in path and path["rect"].intersects(clip_rect):  
                    for item in path["items"]:
                        if item[0] == "l":  
                            shape.draw_line(item[1], item[2])
                        elif item[0] == "re":  
                            shape.draw_rect(item[1])
                        elif item[0] == "qu":  
                            shape.draw_quad(item[1])
                        elif item[0] == "c":  
                            shape.draw_bezier(item[1], item[2], item[3], item[4])

                # Ensure correct color and opacity are applied
                shape.finish(
                    fill=path.get("fill", (1, 1, 1)),  
                    color=path.get("color", (0, 0, 0)),  
                    dashes=path.get("dashes", None),
                    even_odd=path.get("even_odd", True),
                    closePath=path.get("closePath", True),
                    width=path.get("width", 1)
                )

        # Commit the shape to the page
        shape.commit()

        # Save the new PDF into an in-memory buffer
        pdf_buffer = BytesIO()
        new_pdf.save(pdf_buffer)
        pdf_buffer.seek(0)

        # Close the documents
        pdf_document.close()
        new_pdf.close()

        # Return the new PDF as a streaming response
        return StreamingResponse(pdf_buffer, media_type="application/pdf", headers={"Content-Disposition": "attachment; filename=clipped_design.pdf"})

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
`

**Comment by Danishkhilji** [comment]
@JorjMcKie can you look at this?

**Reply by JorjMcKie** [reply]
There is no support for this at the moment. It would need to implement the notion of "clipped areas" in the `Shape` logic to do this properly.
This is currently out of scope.

**Reply by Danishkhilji** [reply]
Thanks, @JorjMcKie for your quick response!
